docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+
+**Note:** The network mesh spanned by OpenVPN uses the network range `192.168.123.0 - 192.168.123.255`. This network range cannot be used in either shoot-, or seed clusters. If it is used this might cause subtle problem due to network range overlaps. Unfortunately, this appears not to be well documented but this restriction exists since the very beginning. We should clean up this technical debt as part of the excercise.
+
+### Goals
+
+* We intend to supersede the current VPN solution with the solution outlined in this proposal.
+* We intend to remove the code for the konnectivity tunnel once this solution proposal has been validated.
+
+### Non Goals
+
+* The solution is not a low latency, or high throughput solution. As the kube-apiserver to shoot cluster traffic does not demand these properties we do not intend to invest in improvements.
+* We do not intend to provide continuous availability to the shoot-seed VPN connection. We expect the availability to be comparable to the existing solution.
+
+## Proposal
+
+The propsal is depicted in the following diagram:
+
+![alt text](assets/ReversedTunnelVPN.png "Reversed Tunnel VPN")
+
+We have added an OpenVPN server pod (`vpn-seed-server`) to each control plane. This OpenVPN server is configured with the OpenVPN `client-to-client` option that allows clients to talk to each other. The OpenVPN client in the shoot cluster (`vpn-shoot-client`) as well as the Open VPN client sidecars in the seed (`vpn-seed-client`) connect to the OpenVPN server.
+
+The three containers `vpn-seed-server`, `vpn-seed-client`, and `vpn-shoot-client` are new containers and are not related to containers in the github.com/gardener/vpn project. We will create a new project github.com/gardener/vpn2 for these containers. With this solution we intend to supersede the containers from the github.com/gardener/vpn project.
+
+A service `openvpn-server` is created for each control plane in its namespace. The `vpn-seed-client` sidecars connect directly to this service.
+
+The `vpn-shoot-client` pod connects to the correct `vpn-seed-server` service via the SNI passthrough proxy introduced with [SNI Passthrough proxy for kube-apiservers](https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md) port 443. No additional port needs to be opened. We re-use the SNI infrastructure in order to avoid the need to have one external load balancer per shoot cluster in the seed.
+
+Shoot OpenVPN clients connect to the correct OpenVPN Server using the http proxy feature provided by OpenVPN. A configuration is added to the envoy proxy to detect http proxy requests and forward the connection attempt to the correct OpenVPN server.
+
+The restriction regarding the `192.268.123.0/24` network range in the current VPN solution still applies to this proposal. No other restrictions are introduced.
```

Comment:
Earlier, you write 

> Unfortunately, this appears not to be well documented but this restriction exists since the very beginning. We should clean up this technical debt as part of the excercise.

What does it mean concretely now?

Reply:
This proposal does not change the fact that the network range 192.168.123.0/24 is off limits for seed and shoot clusters. If it is used nevertheless, the IP space overlap will likely cause hard to debug issues, i.e. packets going to the wrong/unintended targets. Unfortunately, this was never properly documented.
There is no change in this regard concerning the proposal.

Reply:
Understood. What do you plan to improve this in the future (you already pointed out this limitation)? I have the feeling that this could be part of the GEP as you are anyways touching it, so it can be improved on the way?

Reply:
I am not aware of any such future plans. From my perspective, it would make a lot of sense to first document the existing restriction (where?). Additionally, it might be beneficial to actually add some checks to prevent clusters from being created with such an IP space overlap with the vpn solution. Last but not least, the second envoy as proposed by Martin might allow more IP space overlap between seed and shoot, but that needs to be verified first.

Reply:
I think you can go right away to the second proposal, i.e., enhance the existing CIDR validation for seeds and shoots, and forbid using this special range.

Reply:
CIDR validation implemented with #3844.